cmd/snap/cmd_quota_test.go

```diff
@@ -181,17 +209,93 @@ func (s *quotaSuite) TestSetQuotaInvalidArgs(c *check.C) {
 		{[]string{"set-quota", "--memory=888X", "foo"}, `cannot parse "888X\": try 'kB' or 'MB'`},
+		{[]string{"set-quota", "--cpu=0", "foo"}, `cannot parse cpu quota string "0"`},
+		{[]string{"set-quota", "--cpu=0x100%", "foo"}, `cannot parse cpu quota string "0x100%"`},
+		{[]string{"set-quota", "--cpu=2x0%", "foo"}, `cannot parse cpu quota string "2x0%"`},
+		{[]string{"set-quota", "--cpu=200", "foo"}, `cannot parse cpu quota string "200"`},
+		{[]string{"set-quota", "--cpu=20D", "foo"}, `cannot parse cpu quota string "20D"`},
+		{[]string{"set-quota", "--cpu=ASD", "foo"}, `cannot parse cpu quota string "ASD"`},
+		{[]string{"set-quota", "--cpu-set=x", "foo"}, `cannot parse CPU set value "x"`},
+		{[]string{"set-quota", "--cpu-set=1:2", "foo"}, `cannot parse CPU set value "1:2"`},
+		{[]string{"set-quota", "--cpu-set=0,-2", "foo"}, `cannot parse CPU set value "-2"`},
+		{[]string{"set-quota", "--cpu=2x101%", "foo"}, `cannot use value 101: cpu quota percentage must be between 1 and 100`},
+		{[]string{"set-quota", "--thread=xxx", "foo"}, `cannot use thread value "xxx"`},
 		// remove-quota command
 		{[]string{"remove-quota"}, "the required argument `<group-name>` was not provided"},
 	} {
 		s.stdout.Reset()
 		s.stderr.Reset()
 
+		// Mock the CGroup version here to test the --cpu-set command
+		restore := main.MockCGroupVersion(func() (int, error) {
+			return 2, nil
+		})
+
 		_, err := main.Parser(main.Client()).ParseArgs(args.args)
-		c.Assert(err, check.ErrorMatches, args.err)
+		c.Check(err, check.ErrorMatches, args.err, check.Commentf("%q", args.args))
+
+		// Restore the cgroup version callback
+		restore()
 	}
 }
 
+func (s *quotaSuite) TestSetQuotaCpuSetFails(c *check.C) {
+	restore := main.MockIsStdinTTY(true)
+	defer restore()
+
+	const postJSON = `{"type": "async", "status-code": 202,"change":"42", "result": []}`
+	fakeHandlerOpts := fakeQuotaGroupPostHandlerOpts{
+		action:        "ensure",
+		body:          postJSON,
+		groupName:     "foo",
+		cpuCount:      2,
+		cpuPercentage: 50,
+	}
+
+	const getJsonTemplate = `{
+		"type": "sync",
+		"status-code": 200,
+		"result": {
+			"group-name":"foo",
+			"constraints": { "memory": %d },
+			"current": { "memory": 500 }
+		}
+	}`
+
+	cpuSetArgs := []string{"set-quota", "--cpu-set=0,-2", "foo"}
+	cpuArgs := []string{"set-quota", "--cpu=2x50%", "foo"}
+
+	// Mock the CGroup version here to test the --cpu-set and --cpu command
+	restoreCGroup := main.MockCGroupVersion(func() (int, error) {
+		return 1, nil
+	})
+
+	// ensure that --cpu-set does not work with cgroup version 1
+	_, err := main.Parser(main.Client()).ParseArgs(cpuSetArgs)
+	c.Assert(err, check.ErrorMatches, `cannot use CPU set with cgroup version 1`)
+
+	s.stdout.Reset()
+	s.stderr.Reset()
+
+	routes := map[string]http.HandlerFunc{
+		"/v2/quotas": makeFakeQuotaPostHandler(
+			c,
+			fakeHandlerOpts,
+		),
+		"/v2/quotas/foo": makeFakeGetQuotaGroupHandler(c, fmt.Sprintf(getJsonTemplate, 1000)),
```

Comment:
(nitpick) maybe something for a followup but the I would love we had a way to also ensure that these fake handlers actually get called. right now they will error if called twice but not if called not at all. If I e.g. replace `snap set-quota` with just `snap quota` nothing in this test will fail even though the `makeFakeQuotaPostHandler` was never called. But this is a bigger topic than just this PR and maybe something we can address later (or postpone entirely if I'm overthiking it).

Reply:
I think you are correct here, it would make sense to also verify that it _is_ actually called. But I believe that is out of scope for this PR as well